Allow manually overriding suite resource IDs

orangejenny opened the pull request:
##### SUMMARY
This is # 5 under Specification in https://github.com/dimagi/commcare-hq/issues/25718
- Adds a new `ResourceOverride` model
- Adds new overrides when linked apps are pulled and their form's unique ids are updated. This is intended to be removed soon in a later stage of when https://github.com/dimagi/commcare-hq/issues/25718, when the form unique id update code is removed.
- runs a migration to add overrides for all pre-existing linked apps, for their form unique ids

@mkangia My intent is that this could be extended to be used for multimedia resources in the media suite, to do the kind of manual overrides we were discussing in [ICDS-530](https://dimagi-dev.atlassian.net/browse/ICDS-530).

fyi @ctsims

orangejenny committed:
Added ResourceOverride and skeleton for ResourceOverrideHelper

orangejenny committed:
Added body to ResourceOverrideHelper.update_suite

orangejenny committed:
Added test skeleton

orangejenny committed:
Added migration for existing linked apps

orangejenny committed:
Added any new ResourceOverrides on linked app pull

orangejenny committed:
Added basic tests

orangejenny committed:
Added test patches and renamed *_xform_resource_overrides functions

orangejenny committed:
Fixed lint errors

esoergel reviewed (approved):
The ID updating logic has felt a bit sketchy to me, very nice to see it being made more robust!

emord commented on corehq/apps/app_manager/tests/test_advanced_suite.py:
```diff
@@ -22,27 +24,28 @@
 from corehq.util.test_utils import flag_enabled
 
 
+@patch('corehq.apps.app_manager.suite_xml.post_process.resources.get_xform_overrides', return_value=[])
```
Its not clear to me why this needs a mock at all. That function should return an empty dict when called if a record doesn't exist. Also the function you're patching returns dicts, but all of these patches return empty lists, which I believe is unintentional

orangejenny replied:
The empty list bit is unintentional, I'll fix that. The patches are to avoid the db call so these tests don't have to switch from `SimpleTestCase` to `TestCase`.

emord replied:
Ah. avoiding testcase does make sense. This is used enough times and is easy enough to mess up that you may want to consider creating a utility in app manager's test utils. We do something similar for UCRs https://github.com/dimagi/commcare-hq/blob/fd9f7aa24f25093683e17a69bb4a14f44d0e15b7/corehq/apps/userreports/tests/utils.py#L115-L117

orangejenny replied:
Wrapped the patch in a function and fixed the return value in 842d51abb84

I did not remove many of the `*args` that I added earlier, and am not planning to unless there's a compelling reason.

emord commented on corehq/apps/app_manager/suite_xml/post_process/resources.py:
```diff
@@ -0,0 +1,61 @@
+class ResourceOverride(models.Model):
+    domain = models.CharField(max_length=255, null=False)
+    app_id = models.CharField(max_length=255, null=False)
+    root_name = models.CharField(max_length=32, null=False)  # matches up with class in suite_xml.xml_models
```
Is this just matching up with one class in that file, or the `ROOT_NAME` of all classes in that file?

orangejenny replied:
Clarified in a8bc6454892 lmk if it still doesn't make sense.

emord commented on corehq/apps/app_manager/suite_xml/post_process/resources.py:
```diff
@@ -14,4 +17,23 @@ class ResourceOverride(models.Model):
+                resource.id = overrides_by_pre_id[resource.id].post_id
+            id_counts[resource.id] += 1
+
+        duplicates = [id for id, count in id_counts.items() if count > 1]
```
(note if changing to `Counter` you could iterate through `most_common`)

orangejenny replied:
d6628eb9187

emord commented on corehq/apps/app_manager/tests/test_suite_resource_overrides.py:
```diff
@@ -0,0 +1,59 @@
+
+    @classmethod
+    def setUp(self):
+        super().setUpClass()
```
I don't know what this is going to do. Should this be setup or setupclass?

orangejenny replied:
92682c18142

orangejenny committed:
Added meta class

orangejenny commented on corehq/apps/app_manager/suite_xml/post_process/resources.py:
```diff
@@ -0,0 +1,60 @@
+    app_id = models.CharField(max_length=255, null=False)
+    root_name = models.CharField(max_length=32, null=False) # matches up with class in suite_xml.xml_models
+    pre_id = models.CharField(max_length=255, null=False)
+    post_id = models.CharField(max_length=255, null=False)
```
Fixed in 2aac5fb6070 - I had added that to the migration manually and forgot about the meta. Also added an `index_together`, thanks for the suggestion.

orangejenny committed:
Replaced id counting with Counter

orangejenny commented on corehq/apps/app_manager/suite_xml/post_process/resources.py:
```diff
@@ -14,4 +17,23 @@ class ResourceOverride(models.Model):
+        for resource in getattr(self.suite, FormResourceContributor.section_name):
+            if resource.id in overrides_by_pre_id:
+                resource.id = overrides_by_pre_id[resource.id].post_id
+            id_counts[resource.id] += 1
```
d6628eb9187

orangejenny committed:
Clarified comment

orangejenny committed:
Fixed setUp call.

orangejenny committed:
Extracted patch into function and renamed *_xform_overrides functions back to *_xform_resource_overrides

orangejenny committed:
Removed erroneous @classmethod from setUp method

emord commented on corehq/apps/app_manager/suite_xml/post_process/resources.py:
```diff
@@ -0,0 +1,66 @@
+
+    class Meta(object):
+        unique_together = ('domain', 'app_id', 'root_name', 'pre_id')
+        index_together = ('domain', 'app_id', 'root_name')
```
This `index_together` is not useful. these three columns are the leading columns of the `unique_together` so any query that would use these index could use the unique index instead

esoergel replied:
I'm unfamiliar with that technique - is that supported in the Django ORM?  How does one query the unique index?  (Or does postgres optimize that type of query automagically?)

emord replied:
There isn't any "technique" here. Unique constraints in postgres are implemented via an index as noted in https://www.postgresql.org/docs/9.6/indexes-unique.html

> PostgreSQL automatically creates a unique index when a unique constraint or primary key is defined for a table

And then as noted in the docs on [multicolumn indexes](https://www.postgresql.org/docs/9.6/indexes-multicolumn.html), an index can technically be used when a query queries any subset of the columns, but is most effective when there are equality constraints on the leftmost columns

> The exact rule is that equality constraints on leading columns, plus any inequality constraints on the first column that does not have an equality constraint, will be used to limit the portion of the index that is scanned

If the order of these columns were ordered differently from each other, then its possible that they could be useful, but because the multicolumn index has the same order as the unique index, then  the unique index is (almost) strictly better in all cases. 

Of course this is all ignoring the fact that this table may very well be small enough that postgres never actually uses the index and performs a sequential scan anyways. This can happen when the table is very small (so overhead of loading the entire table in memory is small) or that query is expected to return a large portion of the table (say on ICDS where there's only 2 or 3 real apps.

esoergel replied:
Nifty, thanks for the explanation!

orangejenny committed:
Removed unnecessary index_together

mkangia commented on corehq/apps/app_manager/suite_xml/post_process/resources.py:
```diff
@@ -0,0 +1,65 @@
+                resource.id = overrides_by_pre_id[resource.id].post_id
+
+        id_counts = Counter(resource.id for resource in resources)
+        duplicates = [key for key, count in id_counts.items() if count > 1]
```
@emord suggested to use `most_common` here. Was just curious how that would work, so I think this would be
```
# fetch most common id
most_common_id_and_count = id_counts.most_common(1)
if most_common_id_and_count and most_common_id_and_count[0][1] > 1:
    raise ResourceOverrideError("Duplicate resource id found: {}".format(most_common_id_and_count[0][0])
```
but guess we want to get all duplicate ids. Unless there is something different that can be done with `Counter` here.

orangejenny replied:
Yeah, that's what the code would look like - I tried it out and didn't think it was any more readable.

esoergel replied:
Also, FWIW, `most_common` sorts elements by their frequency, but it does also include elements which aren't duplicated.  `counts.most_common(3)` returns the 3 most common elements.  I don't think `Counter` has anything which returns only duplicates, so I like how Jenny's done here, which is simplest.

mkangia replied:
The small snippet i wrote was getting the most common element and if its occurrence is more than 1, that means it was a duplicate.
Agree with what has been done here is simple and cool

mkangia commented:
@orangejenny 
> My intent is that this could be extended to be used for multimedia resources in the media suite, to do the kind of manual overrides we were discussing in ICDS-530.

Can you drop a short note on what manual mm overrides we talked about?

mkangia commented:
Build failure is related
```
corehq.apps.linked_domain.tests.test_linked_apps:TestLinkedApps.test_pull_app_resource_overrides ... ERROR
======================================================================
ERROR: corehq.apps.linked_domain.tests.test_linked_apps:TestLinkedApps.test_pull_app_resource_overrides
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/linked_domain/tests/test_linked_apps.py", line 149, in test_pull_app_resource_overrides
    self._make_build(master_app, True)
  File "/mnt/commcare-hq-ro/corehq/apps/linked_domain/tests/test_linked_apps.py", line 100, in _make_build
    copy = app.make_build()
  File "/mnt/commcare-hq-ro/corehq/util/timer.py", line 223, in _inner
    return fn(self, *args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/models.py", line 4410, in make_build
    copy.create_build_files()
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/models.py", line 4257, in create_build_files
    all_files = self.create_all_files(build_profile_id)
  File "/mnt/commcare-hq-ro/corehq/util/timer.py", line 223, in _inner
    return fn(self, *args, **kwargs)
  File "/vendor/lib/python3.6/site-packages/memoized.py", line 18, in _inner
    return m(*args, **kwargs)
  File "/vendor/lib/python3.6/site-packages/memoized.py", line 140, in __call__
    value = self.func( *args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/models.py", line 5057, in create_all_files
    files.update(self._get_form_files(prefix, build_profile_id))
  File "/mnt/commcare-hq-ro/corehq/util/timer.py", line 223, in _inner
    return fn(self, *args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/models.py", line 5004, in _get_form_files
    raise XFormException(_('Unable to validate the forms due to a server error. '
corehq.apps.app_manager.exceptions.XFormException: Unable to validate the forms due to a server error. Please try again later.
-------------------- >> begin captured logging << --------------------
notify: ERROR: Notify Exception: Error calling Formplayer form validation endpoint
Traceback (most recent call last):
  File "/vendor/lib/python3.6/site-packages/memoized.py", line 138, in __call__
    return cache[key]
KeyError: (None,)
```

orangejenny committed:
Added patch to test

orangejenny commented:
@mkangia 

On ICDS-530, we'd talked about some builds having artificially inflated versions - like, you add an image in v100, then revert to a different stream without the image, make changes, revert back, and you could end up with the image being marked as v150 even though it hasn't actually changed since v100, leading to unnecessary downloads. If we had wanted to address that, at the time I was thinking we'd need to go into old builds and update the build's `multimedia_map` (and then re-run `create_all_files` to overwrite cached build files).

With these overrides, we could extend the model to include version and then manually add an override that would change the version of that image to 100 for any builds that used version 150. We'd still have to re-create cached build files, but it'd be a little less dicey than manipulating the app doc for every build that contained v150. I haven't thought through all of the details, but I think it'd work.

mkangia commented:
Thank you @orangejenny 
Won't this all be taken care of when we roll out multimaster?

orangejenny commented:
@mkangia I don't know what you mean by taken care of, can you say more about that? (or offline with me)

orangejenny commented:
@mkangia Actually, maybe this'll answer: We shouldn't get into the same situation as ICDS-530 again once multimaster is out, I'm just figuring that resource ids & versions are complex enough that there's some chance we'll run into some other issue in the future.

orangejenny committed:
Added arg for patch

mkangia commented:
thanks @orangejenny 
ya i was implying about ICDS-530 not happening again after multimaster release.

> we'll run into some other issue in the future
hopefully not. Do you plan to keep these code changes forever just in case we need it?

orangejenny commented:
@mkangia Hopefully not, indeed.  Additional code to support multimedia isn't necessary until we run into a problem. But most of the changes in this PR are necessary forever, because we're always going to need to override form unique ids in current linked apps, so that those ids don't all change when I remove [this code](https://github.com/dimagi/commcare-hq/blob/774835bd40f42bf28adc14271ac7a77e16a8f355/corehq/apps/app_manager/views/utils.py#L189-L192).

orangejenny commented:
@mkangia I suppose we might be able to remove this code once ~all mobile users are on a CC version that doesn't have the mobile problem with crashing when a form id changes...but that's not going to be for a while.

orangejenny committed:
Added logic to skip migrating builds without an upstream app

orangejenny committed:
Extracted migration to management command so it can be re-run if needed

orangejenny committed:
Added caching to get_xform_resource_overrides

orangejenny committed:
Fixed lint errors

orangejenny committed:
Fixed whitespace

orangejenny committed:
Merge branch 'master' into jls/resource-overrides

orangejenny committed:
Added more info to ResourceOverrideError

orangejenny committed:
Renamed migration after merging master

emord commented on corehq/apps/app_manager/suite_xml/post_process/resources.py:
```diff
@@ -23,18 +23,20 @@ class Meta(object):
+                pre_id=pre_id,
+                post_id=post_id,
             )
             override.save()
```
What happens when this gets saved, but a later value in `pre_to_post_map.items()` raises a `ResourceOverrideError`? Should this be using a transaction over the entire for loop so that this is only committed if there are no errors?

orangejenny replied:
8fb2505d205

emord replied:
`ResourceOverride.objects.create(...)` still creates the object in the loop. This extra save is basically a no-op. I think what you're looking for is:

```
new_overrides = []
for thing in things:
    if should_override:
        new_overrides.append(ResourceOverride(params))

if new_overrides:
    ResourceOverride.objects.bulk_create(new_overrides)
```

orangejenny replied:
thank you - 794b78588f3

emord commented on corehq/apps/app_manager/suite_xml/post_process/resources.py:
```diff
@@ -23,18 +23,20 @@ class Meta(object):
+            if post_id != overrides_by_pre_id[pre_id].post_id:
+                raise ResourceOverrideError("""
+                    Cannot change override: domain {}, app {}, pre_id {}
+                """.strip().format(domain, app_id, pre_id))
```
I think this will only file one error at a time, would it be useful to restructure this so that all errors are returned?

orangejenny replied:
Yeah, I'm reworking this now after running into errors locally. Errors in this are pretty bad news - it'd mean a master form id can be mapped to multiple linked form ids, which means the app is subject to the bug that causes crashes on mobile pre-2.48 - but it's possible that there are apps in this state. I'm likely going to take the approach of writing a verification script and running that before deploying this.

orangejenny replied:
I read this too quickly and thought it was a comment on the migration. Anyway, 8fb2505d205 pretty much deals with this.

orangejenny committed:
Added logic to batch errors in add_xform_resource_overrides

orangejenny committed:
Replaced creates with bulk_create

orangejenny committed:
Added more information to ResourceOverrideError

orangejenny committed:
Added more patches in test_linked_apps.

Tests are intermittently failing with problems with resource overrides being written,
presumably because some of these tests share apps. Avoid this by patching the method
to add overrides, except in the case where it's deliberately being tested, which is
test_pull_app_resource_overrides. The new patch_add_xform_resource_overrides can be
removed when https://github.com/dimagi/commcare-hq/pull/25998 is merged.
